### DISCUSSION TITLE
About effective text length measuring for PowerShell

### DISCUSSION THREAD

**Discussion by lewis-yeung** [JanDeDobbeleer/oh-my-posh#2083] [Ideas]
The current `(*color.Ansi).MeasureText` method uses `utf8.RuneCountInString` to measure the effective text length, it calculates in unit of Unicode characters instead of cells on terminal screen, which will cause tooltips or any right-aligned segments _**in PowerShell**_ to be displayed incorrectly (typically a line break) when the text includes wide characters (e.g., CJK characters) or something like backspaces (mostly act as part of a trick to tweak the rendering results). See the screenshot below:

![screenshot](https://user-images.githubusercontent.com/83903009/163158969-27db74e8-54a5-4ef1-9abf-75a2069f4756.png)

Should we find a better way to handle this?

**Comment by JanDeDobbeleer** [comment]
Up for a good suggestion. It's a tricky thing and requires skills/knowledge I don't possess yet. Might need to have a read about the topic.

**Reply by lewis-yeung** [reply]
Yeah, that needs hard work but is worth trying. Besides, to find a proper solution can involve several topics: character width difference between specific fonts, the way the console/terminal handles tabs (`\t`), emoji rendering width, etc. Now I'm trying to dive into it. 😆

**Reply by JanDeDobbeleer** [reply]
@lewis-yeung keep me in the loop. It's been on my list for quite a while 😃

**Comment by lewis-yeung** [comment]
Ah oh, sorry that I forgot the prerequisite "_**in PowerShell**_". As we know, zsh has a native right prompt support (`$RPROMPT`) so OMP doesn't have to deal with this case for a tooltip. Maybe we can find out how zsh implements it exactly and try to make an implementation in Go or PowerShell.

**Reply by JanDeDobbeleer** [reply]
@lewis-yeung that's shell functionality, not something we need to do. PSReadLine is who you need to talk to for that one.

**Reply by lewis-yeung** [reply]
Aha, you are quite right! I almost forgot that it would make things more and more hacky. 😆

**Comment by lewis-yeung** [comment]
@JanDeDobbeleer Now I come back to this topic again. I just found that [go-runewidth](https://github.com/mattn/go-runewidth) handles most of the wide characters (especially CJK) properly. Although this is not an absolute perfect solution (e.g., the width of a flag emoji cannot be measured correctly for the time being), it does much better than `utf8.RuneCountInString`. I'm ready to submit a PR.

![sample](https://user-images.githubusercontent.com/83903009/177040153-9899c494-f1c9-4e69-be1a-d942e2e54279.png)

**Reply by JanDeDobbeleer** [reply]
@lewis-yeung that would be perfect! That way I don't have to implement custom logic for it.
